src/com/palmergames/bukkit/towny/command/NationCommand.java

```diff
@@ -188,43 +204,110 @@
 	@Override
 	public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
 
-		if (args.length == 1) {
-			List<String> nationNames = NameUtil.getNationNames();
-			nationNames.addAll(nationTabCompletes);
-			return NameUtil.filterByStart(nationNames, args[0]);
-		}
-
-		if (args.length == 2) {
-			switch (args[0].toLowerCase()) {
-				case "set":
-					return NameUtil.filterByStart(nationSetTabCompletes, args[1]);
-				case "rank":
-				case "ally":
-				case "enemy":
-					return NameUtil.filterByStart(new ArrayList<>(Arrays.asList(
-						"add",
-						"remove"
-					)), args[1]);
-				case "kick":
-				case "add":
-					return NameUtil.getTownNamesStartingWith(args[1]);
-				case "toggle":
-					return NameUtil.filterByStart(nationToggleTabCompletes, args[1]);
-				case "join":
-					return NameUtil.getNationNamesStartingWith(args[1]);
-
+		if (sender instanceof Player) {
+			Player player = (Player) sender;
+			switch (args.length) {
+				case 1:
+					// Only suggest nation names in tab complete if the player is not typing a valid sub command. This reduces clutter and the unnecessary checking of all nation names for tab completion
+					List<String> nationNames = NameUtil.filterByStart(nationTabCompletes, args[0]);
+					if (nationNames.size() > 0) {
+						return nationNames;
+					} else {
+						return NameUtil.getNationNamesStartingWith(args[0]);
+					}
+				case 2:
+					switch (args[0].toLowerCase()) {
+						case "set":
+							return NameUtil.filterByStart(nationSetTabCompletes, args[1]);
+						case "rank":
+						case "enemy":
+							return NameUtil.filterByStart(nationEnemyTabCompletes, args[1]);
+						case "ally":
+							return NameUtil.filterByStart(nationAllyTabCompletes, args[1]);
+						case "toggle":
+							return NameUtil.filterByStart(nationToggleTabCompletes, args[1]);
+						case "king":

```

Comment:
Not sure what the king bit here is for, /n set king is a command but I don't think /n king is.

Reply:
`/n king` is not a command but `/n king ?` shows a help page for commands relevant to a nation's king. This isn't necessary but is on purpose.

Reply:
Ah right